### PR TITLE
Immutability/equals/hashCode support for statements

### PR DESCRIPTION
**Pull request by badgerwithagun**
Next step for #7.

Follows on from #43.

This does the same thing for statements as I did previously (in #43) for the lower level elements.  Also introduces a lot more test saturation so we can be sure the assumptions this bakes in continue to be maintained long term.

Once this is in, the next steps are to deprecate and remove all the various cases where mutability is required by existing code so we can start removing some of the mutability support code.

### CODE REVIEW

File: morf-core/src/main/java/org/alfasoftware/morf/sql/AbstractSelectStatement.java [hunk 1]
```diff
@@ -125,6 +161,9 @@ protected void addFields(AliasedFieldBuilder... aliasedFields) {
    * @param aliasedFields The fields to add
    */
   protected void addFields(List<? extends AliasedFieldBuilder> aliasedFields) {
+    if (AliasedField.immutableDslEnabled()) {
```
Comment (tsg21) [inline morf-core/src/main/java/org/alfasoftware/morf/sql/AbstractSelectStatement.java]: Consider pushing this repeated block into a new method on AliasedField: assertImmutableDslEnabled()

File: morf-core/src/main/java/org/alfasoftware/morf/sql/AbstractSelectStatementBuilder.java [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+
+/**
+ * Base builder for select statements.
+ *
+ * @author Copyright (c) Alfa Financial Software 2017
+ *
+ * @param <U> The statement type.
+ * @param <T> The builder type.
+ */
+public abstract class AbstractSelectStatementBuilder<U extends AbstractSelectStatement<U>, T extends AbstractSelectStatementBuilder<U, T>> implements Builder<U> {
```
Comment (tsg21) [inline morf-core/src/main/java/org/alfasoftware/morf/sql/AbstractSelectStatementBuilder.java]: Why is this abstract?

File: morf-core/src/main/java/org/alfasoftware/morf/sql/SelectStatement.java [hunk 3]
```diff
@@ -42,77 +44,104 @@
...
+ *    .from([table])                       = SELECT [fields] FROM [table]
+ *    .innerJoin([table], [criterion])     = SELECT [fields] FROM [table] INNER JOIN [table] ON [criterion]
+ *    .leftOuterJoin(...)
+ *    .where([criterion])                  = SELECT [fields] FROM [table] WHERE [criterion]
+ *    .orderBy([fields])                   = SELECT [fields] FROM [table] ORDER BY [fields]
+ *    .groupBy([fields])                   = SELECT [fields] FROM [table] GROUP BY [fields]
+ *      .having([criterion])               = SELECT [fields] FROM [table] GROUP BY [fields] HAVING [criterion]
+ *    .union([SelectStatement])            = SELECT [fields] FROM [table] UNION [SelectStatement]
+ *    .unionAll([SelectStatement])         = SELECT [fields] FROM [table] UNION ALL [SelectStatement]
+ *    .build()</pre></blockquote>
```
Comment (tsg21) [inline morf-core/src/main/java/org/alfasoftware/morf/sql/SelectStatement.java]: Hmmm, hasn't this reformatting lost some of the flow? It wasn't supposed to imply that each call was chained, but now it looks like they are.

File: morf-core/src/main/java/org/alfasoftware/morf/sql/element/Criterion.java [hunk 4]
```diff
@@ -554,23 +554,11 @@ public String toString() {
...
-      .dispatch(getSelectStatement())
-      .dispatch(getCriteria());
-  }
-
-
-  @SuppressWarnings("rawtypes")
-  private void visitValue(ObjectTreeTraverser traverser,Object value) {
-    if (value instanceof AliasedField) {
-      traverser.dispatch( (AliasedField)value);
-    } else if (value instanceof Iterable) {
```
Comment (tsg21) [inline morf-core/src/main/java/org/alfasoftware/morf/sql/element/Criterion.java]: We've lost this Iterable handling. Not required any more?

File: morf-core/src/main/java/org/alfasoftware/morf/sql/element/Join.java [hunk 5]
```diff
@@ -187,6 +187,9 @@ public SelectStatement getSubSelect() {
    */
   @Deprecated
   public Join on(Criterion onCondition) {
+    if (AliasedField.immutableDslEnabled()) {
```
Comment (tsg21) [inline morf-core/src/main/java/org/alfasoftware/morf/sql/element/Join.java]: This block again...

File: morf-core/src/main/java/org/alfasoftware/morf/sql/AbstractSelectStatement.java (hunk 1 above)
Comment (badgerwithagun) [reply morf-core/src/main/java/org/alfasoftware/morf/sql/AbstractSelectStatement.java]: Couldn't find it repeated anywhere, but moved it there anyway.

File: morf-core/src/main/java/org/alfasoftware/morf/sql/AbstractSelectStatementBuilder.java (hunk 2 above)
Comment (badgerwithagun) [reply morf-core/src/main/java/org/alfasoftware/morf/sql/AbstractSelectStatementBuilder.java]: How could it not be?

File: morf-core/src/main/java/org/alfasoftware/morf/sql/SelectStatement.java (hunk 3 above)
Comment (badgerwithagun) [reply morf-core/src/main/java/org/alfasoftware/morf/sql/SelectStatement.java]: Not sure I follow?

File: morf-core/src/main/java/org/alfasoftware/morf/sql/element/Criterion.java (hunk 4 above)
Comment (badgerwithagun) [reply morf-core/src/main/java/org/alfasoftware/morf/sql/element/Criterion.java]: It's now handled by the traverser (always should have been).  `visitValue(transfers, getValue())` has become `traverser.dispatch(value)`.  That handles and traverses through iterables automatically.

File: morf-core/src/main/java/org/alfasoftware/morf/sql/element/Join.java (hunk 5 above)
Comment (badgerwithagun) [reply morf-core/src/main/java/org/alfasoftware/morf/sql/element/Join.java]: Done

**Comment by badgerwithagun** [comment]
Rebased over master ready for merge

**Comment by badgerwithagun** [comment]
Internal QA complete

**Pull request merged**